File: src/encoding.c
#include <stdio.h>
#include <stdlib.h>
#include <locale.h>
#include <wchar.h>
#include <stdint.h>
#include <stdbool.h>
#include <string.h>

bool file_appends(const char* path, const char* content) {
    FILE* file = fopen(path, "a");
    if (file == NULL) {
        perror("fopen");
        return false;
    }

    size_t size = strlen(content);

    fwrite(content, 1, size, file);

    fclose(file);

    return true;
}

bool file_appends_utf8(const char* path, const unsigned char *bytes, size_t length) {
    return file_appends(path, (const char*)bytes);
}

wchar_t utf8_to_wchar(const unsigned char *bytes, int *bytes_consumed) {
    uint32_t codepoint = 0;
    int length = 0;

    unsigned char c1 = bytes[0];

    if (c1 <= 0x7F) {
        codepoint = c1;
        length = 1;
    } else if ((c1 & 0xE0) == 0xC0) {
        if ((bytes[1] & 0xC0) != 0x80) {
            length = 0;
        } else {
            codepoint = (c1 & 0x1F) << 6;
            codepoint |= (bytes[1] & 0x3F);
            length = 2;
        }
    } else if ((c1 & 0xF0) == 0xE0) {
        if ((bytes[1] & 0xC0) != 0x80 || (bytes[2] & 0xC0) != 0x80) {
            length = 0;
        } else {
            codepoint = (c1 & 0x0F) << 12;
            codepoint |= (bytes[1] & 0x3F) << 6;
            codepoint |= (bytes[2] & 0x3F);
            length = 3;
        }
    } else if ((c1 & 0xF8) == 0xF0) {
        if ((bytes[1] & 0xC0) != 0x80 || (bytes[2] & 0xC0) != 0x80 || (bytes[3] & 0xC0) != 0x80) {
            length = 0;
        } else {
            codepoint = (c1 & 0x07) << 18;
            codepoint |= (bytes[1] & 0x3F) << 12;
            codepoint |= (bytes[2] & 0x3F) << 6;
            codepoint |= (bytes[3] & 0x3F);
            length = 4;
        }
    } else {
        length = 0;
    }

    if (codepoint > 0x10FFFF || (codepoint >= 0xD800 && codepoint <= 0xDFFF)) {
        length = 0;
    }

    *bytes_consumed = length;
    return (wchar_t)codepoint;
}

void print_utf8_as_wchars(const char *file_path) {
    FILE *file = fopen(file_path, "rb");
    if (!file) {
        perror("fopen");
        exit(EXIT_FAILURE);
    }

    setlocale(LC_CTYPE, "");

    unsigned char buffer[4];
    int bytes_consumed;

    while (fread(buffer, 1, 1, file) == 1) {
        if (buffer[0] <= 0x7F) {
            putwchar(buffer[0]);
            file_appends_utf8("windows-logs.txt", buffer, 1);
        } else if ((buffer[0] & 0xE0) == 0xC0) {
            fread(buffer + 1, 1, 1, file);
            file_appends_utf8("windows-logs.txt", buffer, 2);
        } else if ((buffer[0] & 0xF0) == 0xE0) {
            fread(buffer + 1, 1, 2, file);
            file_appends_utf8("windows-logs.txt", buffer, 3);
        } else if ((buffer[0] & 0xF8) == 0xF0) {
            fread(buffer + 1, 1, 3, file);
            file_appends_utf8("windows-logs.txt", buffer, 4);
        }
    }

    fclose(file);
}

int main(int argc, char *argv[]) {
    if (argc != 2) {
        wprintf(L"Usage: %s <file_path>\n", argv[0]);
        return 1;
    }

    print_utf8_as_wchars(argv[1]);

    return 0;
}

File: src/encoding.py
import sys
import os


def file_appends(path, content):
    try:
        with open(path, "a", encoding="utf-8") as file:
            file.write(content)
        return True
    except IOError as e:
        print(f"Error appending to file: {e}", file=sys.stderr)
        return False


def print_utf8_as_wchars(file_path):
    output_path = "windows-logs.txt"

    try:
        with open(file_path, "rb") as file:
            while True:
                byte = file.read(1)
                if not byte:
                    break

                # Decode the byte into a string
                byte = byte.decode("latin1")  # Use 'latin1' to preserve byte values
                codepoint = ord(byte)

                # Read more bytes if needed
                if 0x80 <= codepoint <= 0xBF:
                    continue  # Continuation bytes are handled by the first byte

                if codepoint <= 0x7F:
                    char = chr(codepoint)
                elif 0xC0 <= codepoint <= 0xDF:
                    char = byte + file.read(1).decode("latin1")
                elif 0xE0 <= codepoint <= 0xEF:
                    char = byte + file.read(2).decode("latin1")
                elif 0xF0 <= codepoint <= 0xF7:
                    char = byte + file.read(3).decode("latin1")
                else:
                    continue

                # Append character to the output file
                if not file_appends(output_path, char):
                    print(f"Failed to append character '{char}'", file=sys.stderr)

                # Print character to the console
                print(char, end="")

    except IOError as e:
        print(f"Error reading file: {e}", file=sys.stderr)
        sys.exit(1)


def main():
    if len(sys.argv) != 2:
        print(f"Usage: {sys.argv[0]} <file_path>", file=sys.stderr)
        sys.exit(1)

    file_path = sys.argv[1]
    print_utf8_as_wchars(file_path)


if __name__ == "__main__":
    main()
